mp4parse/src/lib.rs

```diff
@@ -821,6 +1033,40 @@ impl<'a, T: Read> BMFFBox<'a, T> {
     }
 }
 
+impl<'a, T: Read + Offset> BMFFBox<'a, T> {
+    /// Check whether the beginning of `extent` is within the bounds of the `BMFFBox`.
+    /// We assume extents to not cross box boundaries. If so, this will cause an error
+    /// in `read_extent`.
+    fn contains_extent(&self, extent: &ExtentRange) -> bool {
+        if self.offset() <= extent.start() {
+            let start_offset = extent.start() - self.offset();
+            start_offset < self.bytes_left()
+        } else {
+            false
+        }
+    }
+
+    /// Read the range specified by `extent` into `buf` or return an error if the range is not
+    /// fully contained within the `BMFFBox`.
+    fn read_extent(&mut self, extent: &ExtentRange, buf: &mut Vec<u8>) -> Result<()> {
+        let start_offset = extent.start().checked_sub(self.offset()).expect("box does not contain extent");
+        skip(self, start_offset)?;
+        match extent {
+            ExtentRange::WithLength(range) => {
+                let len = range.end.checked_sub(range.start).expect("range start > end");
+                if len > self.bytes_left() {
+                    return Err(Error::InvalidData("extent crosses box boundary"));
+                }
+                self.take(len).read_to_end(buf)?;
```

Comment:
We'll need to make the extension of `buf` here by `read_to_end` use the fallible allocator, somehow.